src/backend/utils/adt/network.c

```diff
@@ -389,6 +416,415 @@ network_cmp(PG_FUNCTION_ARGS)
+ * the heap to compare.
+ *
+ * All inet values have three major components (and take for example the
+ * address 1.2.3.4/24):
+ *
+ *     * Their netmasked bits (1.2.3.0).
+ *     * Their netmask size (/24).
+ *     * Their subnet bits (0.0.0.4).
+ *
+ * cidr values are the same except that with only the first two components --
+ * all their subnet bits *must* be zero (1.2.3.0/24).
+ *
+ * IPv4 and IPv6 are identical in this makeup, with the difference being that
+ * IPv4 addresses have a maximum of 32 bits compared to IPv6's 64 bits, so in
+ * IPv6 each part may be larger.
+ *
+ * inet/cdir types compare using these sorting rules. If any rule is a tie, the
+ * algorithm drops through to the next to break it:
+ *
+ *     1. IPv4 always appears before IPv6.
+ *     2. Netmasked bits are compared.
+ *     3. Netmask size is compared.
+ *     4. All bits are compared (but by this point we know that the netmasked
+ *        bits are equal, so we're in effect only comparing subnet bits).
+ *
+ * When generating abbreviated keys for SortSupport, we pack in as much
+ * information as we have remove for that will ensure that when comparing those
+ * keys as integers, the rules above will be respected.
+ *
+ * In most cases, that means just the most significant of the netmasked bits
+ * are retained, but in the case of IPv4 addresses on 64-bit architecture, we
+ * can hold almost all relevant information for comparisons to the degree that
+ * we'd almost never have to fall back to non-abbreviated comparison. In all
+ * cases, there should be enough data present for the abbreviated keys to
+ * perform very well in all except the very worst of circumstances (and in
+ * those, key abbreviation will probably be aborted as we detect low
+ * cardinality).
+ *
+ * IPv4
+ * ----
+ *
+ * 32-bit architecture:
+ *
+ * Start with 1 bit for the IP family (IPv4 or IPv6; this bit is present in
+ * every case below) followed by all but 1 of the netmasked bits. If those are
+ * all equal, the system will have to fall back to non-abbreviated comparison.
+ *
+ * +----------+---------------------+
+ * | 1 bit IP |   31 bits netmask   |     (up to 1 bit
+ * |  family  |     (truncated)     |   netmask omitted)
+ * +----------+---------------------+
+ *
+ * 64-bit architecture:
+ *
+ * The most interesting case: we have space to store all netmasked bits,
+ * followed by the netmask size, followed by 25 bits of the subnet. We lay the
+ * bits out carefully this way so that the entire value can be compared as an
+ * integer while still adhering to the inet/cidr sorting rules stated above.
+ *
+ * 25 bits is more than enough to store most subnets (all but /6 or smaller),
+ * so in the vast majority of cases these keys should hold enough information
+ * to save trips to the heap.
+ *
+ * +----------+-----------------------+--------------+--------------------+
+ * | 1 bit IP |    32 bits netmask    |    6 bits    |   25 bits subnet   |
```

Comment:
"Up to 25 bits for subnet", or something

Reply:
Alright, changed all these to read "up to 32 bits netmask" and the like.